### PR TITLE
ZEPPELIN-3254. Enable one build of zeppelin to work with multiple versions of spark

### PR DESCRIPTION
**Pull request by zjffdu**
### What is this PR for?
This PR is trying to add test for one build of zeppelin to work with multiple versions of spark. There's 2 main system test
* `SparkIntegrationTest` ( It launches spark interpreter from interpreter setting component in 3 modes: local, yarn-client, yarn-cluster)
* `ZeppelinSparkClusterTest` (It launch spark interpreter in local mode from Paragraph side

### What type of PR is it?
[Improvement]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-3254

### How should this be tested?
* CI pass

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by zjffdu** [comment]
@Leemoonsoo @felixcheung Could you help review it ? Thanks

File: .travis.yml [hunk 1]
```diff
@@ -78,32 +78,46 @@ matrix:
...
+    - sudo: required
+      jdk: "oraclejdk8"
+      dist: trusty
+      env: PYTHON="2" PROFILE="-Pspark-2.2" SPARKR="true" BUILD_FLAG="package -DskipTests -DskipRat -am" TEST_FLAG="test -DskipRat -am" MODULES="-pl zeppelin-zengine,zeppelin-server,spark/interpreter,spark/spark-dependencies" TEST_PROJECTS="-Dtest=ZeppelinSparkClusterTest -DfailIfNoTests=false"
+
+    # Test spark interpreter with different spark versions under python3, only run SparkIntegrationTest
+    - sudo: required
+      jdk: "oraclejdk8"
+      dist: trusty
+      env: PYTHON="3" PROFILE="-Pspark-2.2" SPARKR="true" BUILD_FLAG="package -DskipTests -DskipRat -am" TEST_FLAG="test -DskipRat -am" MODULES="-pl zeppelin-zengine,spark/interpreter,spark/spark-dependencies" TEST_PROJECTS="-Dtest=SparkIntegrationTest -DfailIfNoTests=false"
```
Comment (felixcheung) [inline .travis.yml]: I'm quite concerned that we are adding more to the test matrix, given how often they fail.
Is there a way to merge this into an existing one?

File: spark/interpreter/src/main/java/org/apache/zeppelin/spark/SparkRInterpreter.java [hunk 2]
```diff
@@ -59,12 +59,18 @@ public void open() throws InterpreterException {
...
       sparkRLibPath = System.getenv("SPARK_HOME") + "/R/lib";
-    } else {
+    } else if (System.getenv("ZEPPELIN_HOME") != null){
+      // embedded mode when SPARK_HOME is not specified
       sparkRLibPath = System.getenv("ZEPPELIN_HOME") + "/interpreter/spark/R/lib";
       // workaround to make sparkr work without SPARK_HOME
       System.setProperty("spark.test.home", System.getenv("ZEPPELIN_HOME") + "/interpreter/spark");
+    } else {
+      // yarn-cluster mode
+      sparkRLibPath = "sparkr";
```
Comment (felixcheung) [inline spark/interpreter/src/main/java/org/apache/zeppelin/spark/SparkRInterpreter.java]: hmm, in yarn cluster mode SPARK_HOME is not set?

File: zeppelin-server/notebook/python.recovery [hunk 3]
```diff
@@ -1 +0,0 @@
-2CZA1DVUG:shared_process	192.168.3.2:55410
```
Comment (felixcheung) [inline zeppelin-server/notebook/python.recovery]: is this intentional, python.recovery and python.recovery.crc file?

File: zeppelin-server/src/test/java/org/apache/zeppelin/rest/ZeppelinSparkClusterTest.java [hunk 4]
```diff
@@ -16,588 +16,630 @@
...
+
+  public void setupSparkInterpreter(String sparkHome) throws InterpreterException {
+    InterpreterSetting sparkIntpSetting = ZeppelinServer.notebook.getInterpreterSettingManager()
+        .getInterpreterSettingByName("spark");
+
+    Map<String, InterpreterProperty> sparkProperties =
+        (Map<String, InterpreterProperty>) sparkIntpSetting.getProperties();
+    LOG.info("SPARK HOME detected " + sparkHome);
+    if (System.getenv("SPARK_MASTER") != null) {
+      sparkProperties.put("master",
```
Comment (felixcheung) [inline zeppelin-server/src/test/java/org/apache/zeppelin/rest/ZeppelinSparkClusterTest.java]: should this be `spark.master `?

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkDownloadUtils.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URL;
+
+public class SparkDownloadUtils {
+  private static Logger LOGGER = LoggerFactory.getLogger(SparkDownloadUtils.class);
+
+  private static String downloadFolder = System.getProperty("user.home") + "/.cache/spark";
```
Comment (felixcheung) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkDownloadUtils.java]: I don't know how much we are supporting Windows nowadays, but we would need to be a different subpath

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkDownloadUtils.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      LOGGER.info("Skip to download spark as it is already downloaded.");
+      return targetSparkHomeFolder.getAbsolutePath();
+    }
+    // Try mirrors a few times until one succeeds
+    for (int i = 0; i < 3; i++) {
+      try {
+        String preferredMirror = IOUtils.toString(new URL("https://www.apache.org/dyn/closer.lua?preferred=true"));
+        File downloadFile = new File(downloadFolder + "/spark-" + version + "-bin-hadoop2.6.tgz");
+        String downloadURL = preferredMirror + "/spark/spark-" + version + "/spark-" + version + "-bin-hadoop2.6.tgz";;
+        runShellCommand(new String[] {"wget", downloadURL, "-P", downloadFolder});
```
Comment (felixcheung) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkDownloadUtils.java]: .. also not on windows..

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkIntegrationTest.java [hunk 7]
```diff
@@ -8,21 +8,47 @@
...
+  public SparkIntegrationTest(String sparkVersion) {
+    LOGGER.info("Testing SparkVersion: " + sparkVersion);
+    this.sparkVersion = sparkVersion;
+    this.sparkHome = SparkDownloadUtils.downloadSpark(sparkVersion);
+  }
+
+  @Parameterized.Parameters
+  public static List<Object[]> data() {
+    return Arrays.asList(new Object[][]{
+        {"2.2.1"},
```
Comment (felixcheung) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkIntegrationTest.java]: add 2.3.0? (or we need another PR for that I guess)

File: .travis.yml (hunk 1 above)
Comment (zjffdu) [reply .travis.yml]: Right, I plan to do that in https://issues.apache.org/jira/browse/ZEPPELIN-3275

File: spark/interpreter/src/main/java/org/apache/zeppelin/spark/SparkRInterpreter.java (hunk 2 above)
Comment (zjffdu) [reply spark/interpreter/src/main/java/org/apache/zeppelin/spark/SparkRInterpreter.java]: There's no SPARK_HOME in driver of yarn cluster mode. As the driver run in yarn container.

File: zeppelin-server/notebook/python.recovery (hunk 3 above)
Comment (zjffdu) [reply zeppelin-server/notebook/python.recovery]: yes, these files are not used, should be removed.

File: zeppelin-server/src/test/java/org/apache/zeppelin/rest/ZeppelinSparkClusterTest.java (hunk 4 above)
Comment (zjffdu) [reply zeppelin-server/src/test/java/org/apache/zeppelin/rest/ZeppelinSparkClusterTest.java]: Zeppelin use both spark.master & master which are the same.

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkIntegrationTest.java (hunk 7 above)
Comment (zjffdu) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkIntegrationTest.java]: Yes, 2.3.0 is just released, will do that in another PR.

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkDownloadUtils.java (hunk 5 above)
Comment (zjffdu) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkDownloadUtils.java]: I am afraid we need to change a lot for windows support besides this, zeppelin doesn't officially support windows IIUC, even we add `.appveyor.yml`, there's no build there actually.

Comment (felixcheung) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkDownloadUtils.java]: well, we used to get a fair number of questions related to running zeppelin on windows in user@ or dev@, I think we should raise this before the next release with these changes...

Comment (zjffdu) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/SparkDownloadUtils.java]: make sense

**Comment by zjffdu** [comment]
Will merge if no more comments

**Pull request closed without merging**